taskmgr/queueservice.go

```diff
@@ -0,0 +1,387 @@
+		// would return an object on the form: {taskId, runId}.
+		m, err := base64.StdEncoding.DecodeString(qm.MessageText)
+		if err != nil {
+			// try to delete from Azure, if it fails, nothing we can do about it
+			// not very serious - another worker will try to delete it
+			q.Log.Errorf("Not able to base64 decode the Message Text '%s' in Azure QueueMessage response.'", qm.MessageText)
+			q.Log.Info("Deleting from Azure queue as other workers will have the same problem.")
+			err := q.deleteFromAzure("", signedDeleteUrl)
+			if err != nil {
+				q.Log.WithFields(logrus.Fields{
+					"url":   signedDeleteUrl,
+					"error": err,
+				}).Warn("Not able to call Azure delete URL")
+			}
+			return nil, err
+		}
+
+		// initialise fields of TaskRun not contained in json string m
+		taskRun := &TaskRun{
+			SignedDeleteUrl: signedDeleteUrl,
+		}
+
+		// now populate remaining json fields of TaskRun from json string m
+		err = json.Unmarshal(m, &taskRun)
+		if err != nil {
+			logger.WithFields(logrus.Fields{
+				"error":   err,
+				"message": m,
+			}).Warn("Not able to unmarshal json from base64 decoded MessageText")
+			err := q.deleteFromAzure("", signedDeleteUrl)
+			if err != nil {
+				logger.WithFields(logrus.Fields{
+					"url":   signedDeleteUrl,
+					"error": err,
+				}).Warn("Not able to call Azure delete URL")
+			}
+			continue
+		}
+		taskRuns = append(taskRuns, taskRun)
+	}
+
+	return taskRuns, nil
+}
+
+// Refreshes a list of task queue urls.  Each task queue contains a pair of signed urls
+// used for polling and deleting messages.
+func (q *queueService) refreshTaskQueueUrls() error {
+	// Attempt to wait until expiration gets closer before refreshing.  No
+	// need to do it more frequently.
+	if !q.shouldRefreshQueueUrls() {
+		return nil
+	}
+
+	q.Log.Debug("Refreshing Azure queue task urls")
+	signedURLs, _, err := q.client.PollTaskUrls(q.ProvisionerId, q.WorkerType)
+	if err != nil {
+		q.Log.WithField("error", err).Warn("Error retrieving task urls.")
+		return errors.New("Error retrieving task urls.")
+	}
+
+	taskQueues := []TaskQueue{}
+	for _, pair := range signedURLs.Queues {
+		taskQueues = append(taskQueues, TaskQueue(pair))
+	}
+
+	q.mu.Lock()
+	q.queues = taskQueues
+	q.Expires = signedURLs.Expires
+	q.mu.Unlock()
+	q.Log.Debugf("Refreshed %d Azure queue task urls", len(taskQueues))
+	return nil
+}
+
+func (q queueService) retrieveTasksFromQueue(ntasks int) (*[]*TaskRun, error) {
+	err := q.refreshTaskQueueUrls()
+	if err != nil {
+		return nil, err
+	}
+
+	tasks := []*TaskRun{}
+	for _, queue := range q.queues {
+		// It hopefully would never be greater, but just incase, we would want to return
+		// and run what tasks we do have.
+		if len(tasks) >= ntasks {
+			return &tasks, nil
+		}
+		taskRuns, err := q.pollTaskUrl(&queue, ntasks-len(tasks), q.Log)
+		if err != nil {
+			q.Log.Warnf("Could not retrieve tasks from the Azure queue. %s", err)
+			continue
+		}
+		tasks = append(tasks, taskRuns...)
```

Comment:
Just outlining what I think is a better interface... Perhaps we can:
  A) implement this interface using the current queue service... it won't have the benefits of (1) - (4), but we'll be able to improve it later (without having to refactor the rest of the worker package)
  B) Not do anything and refactor this later...